File: packages/core/src/utils/series-canon.ts
/**
 * R20 系列正典共享（389 号契约批，四轮 P0；v4 §3 R20）。
 *
 * Sudowrite Series Folders 对标：跨书共享的正典层——同系列多书消费
 * 同一套角色/设定条目，作者一次编写、逐书命中注入。核心契约：
 *
 * 1. `parseSeriesCanonFile`：`.inkos/series/{seriesId}.json` 解析
 *    （version=1 整包校验；条目逐条校验非法跳过；产物按 name 码元序）；
 * 2. `mergeCodexLayers`：双层合并——book 同名条目**覆盖** series 条目
 *    （覆盖即省略：幸存 series 条目单独注入，book 块零改动）；
 * 3. `renderSeriesCodexBlock`：幸存条目命中 → 「## 系列实体卡」块（双语）。
 *
 * 条目与 EntityCodexCard 同构（kind 沿用名册五枚举，事件类条目落
 * "other"）；seriesId 为 snake_case slug（禁空格/连字符，≤64 码元）。
 *
 * 双端：`engine-rs/src/utils/series_canon.rs` 1:1 镜像；共享向量
 * `src/__tests__/golden/series-canon-vectors.json`（TS 断言
 * `golden-series-canon.test.ts`；Rust 差分 `tests/golden_series_canon_diff.rs`）。
 */

import { CODEX_LIMITS, type CodexMatch, type EntityCodexCard } from "./entity-codex.js";

export const SERIES_CANON_VERSION = 1 as const;

export interface SeriesCanonFile {
  readonly version: typeof SERIES_CANON_VERSION;
  readonly seriesId: string;
  readonly entries: ReadonlyArray<EntityCodexCard>;
}

const SERIES_ID_PATTERN = /^[a-z0-9_]{1,64}$/;

/** seriesId 规范：snake_case slug（禁空格/连字符/路径段，≤64 码元）。 */
export function isValidSeriesId(seriesId: string): boolean {
  return SERIES_ID_PATTERN.test(seriesId);
}

const CODEX_KINDS = new Set(["person", "place", "faction", "item", "other"]);
const SERIES_LIMITS = { name: 80, aliases: 10, alias: 60 } as const;

const clampText = (value: string, maxChars: number): string =>
  value.length > maxChars ? value.slice(0, maxChars) : value;

const byName = (a: EntityCodexCard, b: EntityCodexCard): number =>
  a.name < b.name ? -1 : a.name > b.name ? 1 : 0;

/** 条目校验（非法跳过/字段截断）；kind 非法兜底 other（对齐名册解析）。 */
function sanitizeEntry(raw: unknown): EntityCodexCard | undefined {
  if (typeof raw !== "object" || raw === null) return undefined;
  const record = raw as Record<string, unknown>;
  const name = typeof record.name === "string" ? record.name.trim() : "";
  if (!name) return undefined;
  const aliases = Array.isArray(record.aliases)
    ? record.aliases
        .map((alias) => (typeof alias === "string" ? alias.trim() : ""))
        .filter(Boolean)
        .slice(0, SERIES_LIMITS.aliases)
        .map((alias) => clampText(alias, SERIES_LIMITS.alias))
    : [];
  const kindRaw = typeof record.kind === "string" ? record.kind.trim() : "";
  const kind = (CODEX_KINDS.has(kindRaw) ? kindRaw : "other") as EntityCodexCard["kind"];
  const summary = typeof record.summary === "string" ? clampText(record.summary, CODEX_LIMITS.summary) : "";
  const facts = Array.isArray(record.facts)
    ? record.facts
        .map((fact) => (typeof fact === "string" ? fact.trim() : ""))
        .filter(Boolean)
        .slice(0, CODEX_LIMITS.facts)
        .map((fact) => clampText(fact, CODEX_LIMITS.fact))
    : [];
  const relationships = Array.isArray(record.relationships)
    ? record.relationships
        .map((relation) => {
          if (typeof relation !== "object" || relation === null) return undefined;
          const rel = relation as Record<string, unknown>;
          const target = typeof rel.target === "string" ? rel.target.trim() : "";
          if (!target) return undefined;
          const note = typeof rel.note === "string" ? clampText(rel.note, CODEX_LIMITS.relationshipNote) : "";
          return { target, note };
        })
        .filter((relation): relation is { target: string; note: string } => relation !== undefined)
        .slice(0, CODEX_LIMITS.relationships)
    : [];
  const firstChapter =
    typeof record.firstChapter === "number" &&
    Number.isInteger(record.firstChapter) &&
    record.firstChapter >= 0
      ? record.firstChapter
      : undefined;
  return {
    name: clampText(name, SERIES_LIMITS.name),
    aliases,
    kind,
    summary,
    facts,
    relationships,
    ...(firstChapter !== undefined ? { firstChapter } : {}),
  };
}

/** 解析系列正典文件（version/seriesId 整包校验；坏条目跳过；name 码元序）。 */
export function parseSeriesCanonFile(raw: unknown): SeriesCanonFile | undefined {
  if (typeof raw !== "object" || raw === null) return undefined;
  const record = raw as Record<string, unknown>;
  if (record.version !== SERIES_CANON_VERSION) return undefined;
  const seriesId = typeof record.seriesId === "string" ? record.seriesId : "";
  if (!isValidSeriesId(seriesId)) return undefined;
  if (!Array.isArray(record.entries)) return undefined;
  const entries = record.entries
    .map(sanitizeEntry)
    .filter((entry): entry is EntityCodexCard => entry !== undefined)
    .sort(byName);
  return { version: SERIES_CANON_VERSION, seriesId, entries };
}

export interface CodexLayerMerge {
  /** 书内卡片（原样返回，顺序不变）。 */
  readonly book: ReadonlyArray<EntityCodexCard>;
  /** 幸存 series 条目（同名被 book 覆盖剔除后，name 码元序）。 */
  readonly series: ReadonlyArray<EntityCodexCard>;
}

/** 双层合并：book 同名条目覆盖 series 条目（覆盖即省略，非交织）。 */
export function mergeCodexLayers(
  bookCards: ReadonlyArray<EntityCodexCard>,
  seriesEntries: ReadonlyArray<EntityCodexCard>,
): CodexLayerMerge {
  const bookNames = new Set(bookCards.map((card) => card.name));
  const series = seriesEntries.filter((entry) => !bookNames.has(entry.name)).sort(byName);
  return { book: bookCards, series };
}

const SERIES_EXCERPT_MAX = 400;

/** 注入渲染：幸存条目命中 → 「## 系列实体卡」块（无命中 undefined）。 */
export function renderSeriesCodexBlock(
  matches: ReadonlyArray<CodexMatch<EntityCodexCard>>,
  language: "zh" | "en" = "zh",
): string | undefined {
  if (matches.length === 0) return undefined;
  const isEn = language === "en";
  const sections = matches.map(({ card }) => {
    const aliasPart = card.aliases.length > 0 ? `（${card.aliases.join("、")}）` : "";
    const lines = [`### ${card.name}${aliasPart}`];
    if (card.summary) {
      lines.push(
        card.summary.length > SERIES_EXCERPT_MAX
          ? `${card.summary.slice(0, SERIES_EXCERPT_MAX)}…`
          : card.summary,
      );
    }
    if (card.facts.length > 0) {
      lines.push(isEn ? "Canon facts:" : "正典事实：");
      for (const fact of card.facts) lines.push(`- ${fact}`);
    }
    if (card.relationships.length > 0) {
      lines.push(isEn ? "Relationships:" : "关系：");
      for (const relation of card.relationships) {
        lines.push(`- ${relation.target}: ${relation.note}`);
      }
    }
    return lines.join("\n");
  });
  return [
    isEn
      ? "## Series entity cards (cross-book canon — follow canon facts)"
      : "## 系列实体卡（跨书正典——遵循正典事实）",
    ...sections,
  ].join("\n");
}

File: engine-rs/src/utils/series_canon.rs
//! R20 系列正典共享（389 号契约批，四轮 P0）。
//!
//! TS 真源：`packages/core/src/utils/series-canon.ts`；golden 唯一事实源：
//! `packages/core/src/__tests__/golden/series-canon-vectors.json`（差分测试
//! `tests/golden_series_canon_diff.rs` 读同一文件）。
//!
//! 跨书正典层（Sudowrite Series Folders 对标）：`.inkos/series/{seriesId}.json`
//! 同系列多书消费；book 同名条目覆盖 series 条目（覆盖即省略）。
//! 截断/计数与 TS 逐字一致（码元 = chars；禁 locale 感知排序）。

use crate::utils::entity_codex::{CodexMatch, EntityCodexCard};
use serde::{Deserialize, Serialize};
use serde_json::Value;

pub const SERIES_CANON_VERSION: i64 = 1;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SeriesCanonFile {
    pub version: i64,
    pub series_id: String,
    pub entries: Vec<EntityCodexCard>,
}

/// seriesId 规范：snake_case slug（禁空格/连字符/路径段，≤64 码元）。
pub fn is_valid_series_id(series_id: &str) -> bool {
    let len = series_id.chars().count();
    (1..=64).contains(&len)
        && series_id
            .chars()
            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_')
}

const SERIES_LIMITS_NAME: usize = 80;
const SERIES_LIMITS_ALIASES: usize = 10;
const SERIES_LIMITS_ALIAS: usize = 60;
const CODEX_LIMITS_SUMMARY: usize = 300;
const CODEX_LIMITS_FACTS: usize = 10;
const CODEX_LIMITS_FACT: usize = 200;
const CODEX_LIMITS_RELATIONSHIPS: usize = 10;
const CODEX_LIMITS_RELATIONSHIP_NOTE: usize = 200;

fn clamp_text(value: &str, max_chars: usize) -> String {
    if value.chars().count() > max_chars {
        value.chars().take(max_chars).collect()
    } else {
        value.to_string()
    }
}

fn clamp_string_list(raw: Option<&Vec<Value>>, max_items: usize, max_chars: usize) -> Vec<String> {
    raw.map(|items| {
        items
            .iter()
            .filter_map(|item| item.as_str())
            .map(|item| item.trim().to_string())
            .filter(|item| !item.is_empty())
            .take(max_items)
            .map(|item| clamp_text(&item, max_chars))
            .collect()
    })
    .unwrap_or_default()
}

fn is_valid_kind(kind: &str) -> bool {
    matches!(kind, "person" | "place" | "faction" | "item" | "other")
}

/// 条目校验（非法跳过/字段截断）；kind 非法兜底 other（对齐名册解析）。
fn sanitize_entry(raw: &Value) -> Option<EntityCodexCard> {
    let record = raw.as_object()?;
    let name = record
        .get("name")
        .and_then(Value::as_str)
        .unwrap_or("")
        .trim()
        .to_string();
    if name.is_empty() {
        return None;
    }
    let name = clamp_text(&name, SERIES_LIMITS_NAME);
    let aliases = clamp_string_list(
        record.get("aliases").and_then(Value::as_array),
        SERIES_LIMITS_ALIASES,
        SERIES_LIMITS_ALIAS,
    );
    let kind_raw = record
        .get("kind")
        .and_then(Value::as_str)
        .unwrap_or("")
        .trim();
    let kind = if is_valid_kind(kind_raw) {
        kind_raw.to_string()
    } else {
        "other".to_string()
    };
    let summary = clamp_text(
        record.get("summary").and_then(Value::as_str).unwrap_or(""),
        CODEX_LIMITS_SUMMARY,
    );
    let facts = clamp_string_list(
        record.get("facts").and_then(Value::as_array),
        CODEX_LIMITS_FACTS,
        CODEX_LIMITS_FACT,
    );
    let relationships: Vec<crate::utils::entity_codex::CodexRelationship> = record
        .get("relationships")
        .and_then(Value::as_array)
        .map(|items| {
            items
                .iter()
                .filter_map(|relation| {
                    let rel = relation.as_object()?;
                    let target = rel.get("target").and_then(Value::as_str)?.trim().to_string();
                    if target.is_empty() {
                        return None;
                    }
                    let note = clamp_text(
                        rel.get("note").and_then(Value::as_str).unwrap_or(""),
                        CODEX_LIMITS_RELATIONSHIP_NOTE,
                    );
                    Some(crate::utils::entity_codex::CodexRelationship { target, note })
                })
                .take(CODEX_LIMITS_RELATIONSHIPS)
                .collect()
        })
        .unwrap_or_default();
    let first_chapter = record
        .get("firstChapter")
        .and_then(Value::as_i64)
        .filter(|chapter| *chapter >= 0);
    Some(EntityCodexCard {
        name,
        aliases,
        kind,
        summary,
        facts,
        relationships,
        first_chapter,
    })
}

/// 解析系列正典文件（version/seriesId 整包校验；坏条目跳过；name 码元序）。
pub fn parse_series_canon_file(raw: &Value) -> Option<SeriesCanonFile> {
    let record = raw.as_object()?;
    if record.get("version").and_then(Value::as_i64) != Some(SERIES_CANON_VERSION) {
        return None;
    }
    let series_id = record.get("seriesId").and_then(Value::as_str)?;
    if !is_valid_series_id(series_id) {
        return None;
    }
    let entries_raw = record.get("entries").and_then(Value::as_array)?;
    let mut entries: Vec<EntityCodexCard> = entries_raw
        .iter()
        .filter_map(sanitize_entry)
        .collect();
    entries.sort_by(|a, b| a.name.cmp(&b.name));
    Some(SeriesCanonFile {
        version: SERIES_CANON_VERSION,
        series_id: series_id.to_string(),
        entries,
    })
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CodexLayerMerge {
    /// 书内卡片（原样返回，顺序不变）。
    pub book: Vec<EntityCodexCard>,
    /// 幸存 series 条目（同名被 book 覆盖剔除后，name 码元序）。
    pub series: Vec<EntityCodexCard>,
}

/// 双层合并：book 同名条目覆盖 series 条目（覆盖即省略，非交织）。
pub fn merge_codex_layers(
    book_cards: &[EntityCodexCard],
    series_entries: &[EntityCodexCard],
) -> CodexLayerMerge {
    let mut series: Vec<EntityCodexCard> = series_entries
        .iter()
        .filter(|entry| !book_cards.iter().any(|card| card.name == entry.name))
        .cloned()
        .collect();
    series.sort_by(|a, b| a.name.cmp(&b.name));
    CodexLayerMerge {
        book: book_cards.to_vec(),
        series,
    }
}

const SERIES_EXCERPT_MAX: usize = 400;

/// 注入渲染：幸存条目命中 → 「## 系列实体卡」块（无命中返回 None）。
pub fn render_series_codex_block(
    matches: &[CodexMatch],
    language: crate::utils::language::WritingLanguage,
) -> Option<String> {
    if matches.is_empty() {
        return None;
    }
    let is_en = language == crate::utils::language::WritingLanguage::En;
    let sections: Vec<String> = matches
        .iter()
        .map(|codex_match| {
            let card = &codex_match.card;
            let mut lines = Vec::new();
            if card.aliases.is_empty() {
                lines.push(format!("### {}", card.name));
            } else {
                lines.push(format!("### {}（{}）", card.name, card.aliases.join("、")));
            }
            if !card.summary.is_empty() {
                if card.summary.chars().count() > SERIES_EXCERPT_MAX {
                    let clipped: String = card.summary.chars().take(SERIES_EXCERPT_MAX).collect();
                    lines.push(format!("{clipped}…"));
                } else {
                    lines.push(card.summary.clone());
                }
            }
            if !card.facts.is_empty() {
                lines.push(if is_en { "Canon facts:".to_string() } else { "正典事实：".to_string() });
                for fact in &card.facts {
                    lines.push(format!("- {fact}"));
                }
            }
            if !card.relationships.is_empty() {
                lines.push(if is_en { "Relationships:".to_string() } else { "关系：".to_string() });
                for relation in &card.relationships {
                    lines.push(format!("- {}: {}", relation.target, relation.note));
                }
            }
            lines.join("\n")
        })
        .collect();
    let header = if is_en {
        "## Series entity cards (cross-book canon — follow canon facts)"
    } else {
        "## 系列实体卡（跨书正典——遵循正典事实）"
    };
    let mut parts = vec![header.to_string()];
    parts.extend(sections);
    Some(parts.join("\n"))
}
